### PR TITLE
Improve 'waiting for' message by supplying a reason for the message

### LINKED ISSUE #2069: "waiting for locks or build slots" is not a very helpful message
I think the message that nix prints when something is blocking the build could be a bit more descriptive. "waiting for locks or build slots" does not indicate whether there is a resource problem that the user needs to go fix (let alone what the resource problem is), or whether the same build is already underway by another process. Even turning the logging up to vomit doens't yield anything particularly insightful.

### PR DESCRIPTION
**Pull request by balsoft**
Resolves #2069 by allowing to pass the reason for which we are waiting to `Worker::wait*` functions.

Implementation details: stores the latest reason for which some `DerivationGoal` is sleeping in `Worker`, and if a `wait*` function is called with a reason different from the previous one, that reason is printed to stderr. Example output:

```
$ nix build -f test-nix.nix --max-jobs 8 --no-substitute
waiting for build slots...
```

### CODE REVIEW

File: src/libstore/build.cc [hunk 1]
```diff
@@ -1464,7 +1469,7 @@ void DerivationGoal::tryToBuild()
             case rpPostpone:
                 /* Not now; wait until at least one child finishes or
                    the wake-up timeout expires. */
-                worker.waitForAWhile(shared_from_this());
+                worker.waitForAWhile(shared_from_this(), "build hook");
```
Comment (kirelagin) [inline src/libstore/build.cc:1472]:
```suggestion
                worker.waitForAWhile(shared_from_this(), "a build hook");
```

File: src/libstore/build.cc [hunk 2]
```diff
@@ -4451,7 +4456,7 @@ void SubstitutionGoal::tryToRun()
        a substituter to run.  This is because substitutions cannot be
        distributed to another machine via the build hook. */
     if (worker.getNrLocalBuilds() >= std::max(1U, (unsigned int) settings.maxBuildJobs)) {
-        worker.waitForBuildSlot(shared_from_this());
+        worker.waitForBuildSlot(shared_from_this(), "substitution slot");
```
Comment (kirelagin) [inline src/libstore/build.cc:4459]:
```suggestion
        worker.waitForBuildSlot(shared_from_this(), "a substitution slot");
```

File: src/libstore/build.cc [hunk 3]
```diff
@@ -4706,27 +4711,36 @@ void Worker::childTerminated(Goal * goal, bool wakeSleepers)
...
+        printError(format("waiting for %1%...") % resourceName);
+    }
+}
 
-void Worker::waitForBuildSlot(GoalPtr goal)
+void Worker::waitForBuildSlot(GoalPtr goal, string resourceName)
 {
-    debug("wait for build slot");
+    debug("wait for build slots");
+    updateBlockingResourceName(resourceName);
```
Comment (kirelagin) [inline src/libstore/build.cc:4724]: I think this line will print the message even if we are not actually waiting, i.e. if we take the first branch in the `if` below. I’m not sure if that is the desired behaviour or not.

Comment (balsoft) [reply src/libstore/build.cc:4724]: Yes, thanks for noticing that. I think the reason I did that is that `debug("wait for build slot")` was here...

**Comment by balsoft** [comment]
No, not exactly. It prints those messages slightly earlier in the code -- right at the time when the waiting is scheduled, whereas previously it was printed before the thread goes to sleep. However, it shouldn't be much of a difference (if I understand the execution path correctly).

**Comment by edolstra** [comment]
Thanks, I've gone for a slightly different approach in 5ed5d7acbd00298df7b4e3a638d731143084da2e (using `Activity` to make "waiting for lock" messages show in the progress bar).

**Comment by balsoft** [comment]
Thanks!

**Pull request closed without merging**